arangod/Replication/TailingSyncer.cpp

```diff
@@ -827,24 +827,41 @@ Result TailingSyncer::truncateCollection(arangodb::velocypack::Slice const& slic
     return Result(TRI_ERROR_REPLICATION_UNEXPECTED_TRANSACTION, msg);
   }
 
-  SingleCollectionTransaction trx(transaction::StandaloneContext::Create(*vocbase),
-                                  *col, AccessMode::Type::EXCLUSIVE);
-  trx.addHint(transaction::Hints::Hint::INTERMEDIATE_COMMITS);
-  trx.addHint(transaction::Hints::Hint::ALLOW_RANGE_DELETE);
-  Result res = trx.begin();
-  if (!res.ok()) {
-    return res;
-  }
+  uint64_t count = 0;
+  Result res;
+  {
+    SingleCollectionTransaction trx(transaction::StandaloneContext::Create(*vocbase),
+                                    *col, AccessMode::Type::EXCLUSIVE);
+    trx.addHint(transaction::Hints::Hint::INTERMEDIATE_COMMITS);
+    trx.addHint(transaction::Hints::Hint::ALLOW_RANGE_DELETE);
+    res = trx.begin();
+    if (!res.ok()) {
+      return res;
+    }
+                
+    OperationResult opRes = trx.count(col->name(), transaction::CountType::Normal);
+    if (opRes.ok() && opRes.slice().isNumber()) {
+      count = opRes.slice().getNumber<uint64_t>();
+    }
 
-  OperationOptions opts;
-  opts.isRestore = true;
-  OperationResult opRes = trx.truncate(col->name(), opts);
+    OperationOptions opts;
+    opts.isRestore = true;
+    opRes = trx.truncate(col->name(), opts);
 
-  if (opRes.fail()) {
-    return opRes.result;
+    if (opRes.fail()) {
+      return opRes.result;
+    }
+
+    res = trx.finish(opRes.result);
+  }
+
+  if (res.ok() && count >= 4 * 1024) {
+    // only compact if the collection contained a substantial amount of documents
+    // before truncation
+    col->compact();
```

Comment:
This code can be found again in `RestCollectionHandler.cpp`. I wonder if a) there are more places that could do a truncate (haven't checked that, just asking) and b.) if there should be a function, which does exactly that, a compacting truncate.